c/secp256k1_blake160_multisig_all.c

```diff
@@ -0,0 +1,220 @@
+    return ERROR_WITNESS_TOO_LONG;
+  }
+  if (len < FLAGS_SIZE) {
+    return ERROR_WITNESS_TOO_SHORT;
+  }
+
+  /* Get flags */
+  uint8_t pubkeys_cnt = witness[3];
+  uint8_t threshold = witness[2];
+  uint8_t require_first_n = witness[1];
+  if (pubkeys_cnt == 0) {
+    return ERROR_INVALID_PUBKEYS_CNT;
+  }
+  if (threshold > pubkeys_cnt) {
+    return ERROR_INVALID_THRESHOLD;
+  }
+  if (threshold == 0) {
+    threshold = pubkeys_cnt;
+  }
+  if (require_first_n > threshold) {
+    return ERROR_INVALID_REQUIRE_FIRST_N;
+  }
+  size_t multisig_script_len = FLAGS_SIZE + PUBKEY_SIZE * pubkeys_cnt;
+  size_t required_witness_len = multisig_script_len + SIGNATURE_SIZE * threshold;
+  if (len < required_witness_len) {
+    return ERROR_WITNESS_TOO_SHORT;
+  }
+  size_t extra_witness_len = len - required_witness_len;
+
+  /* Check multisig script hash */
+  blake2b_state blake2b_ctx;
+  blake2b_init(&blake2b_ctx, BLAKE2B_BLOCK_SIZE);
+  blake2b_update(&blake2b_ctx, witness, multisig_script_len);
+  blake2b_final(&blake2b_ctx, temp, BLAKE2B_BLOCK_SIZE);
+
+  if (memcmp(bytes_res.pos.ptr, temp, BLAKE160_SIZE) != 0) {
+    return ERROR_MULTSIG_SCRIPT_HASH;
+  }
+
+  /* Load tx hash */
+  unsigned char tx_hash[BLAKE2B_BLOCK_SIZE];
+  len = BLAKE2B_BLOCK_SIZE;
+  ret = ckb_load_tx_hash(tx_hash, &len, 0);
+  if (ret != CKB_SUCCESS) {
+    return ERROR_SYSCALL;
+  }
+
+  /* Verify threshold signatures */
+  uint8_t used_signatures[threshold];
+  memset(used_signatures, 0, threshold);
+
+  secp256k1_context context;
+  uint8_t secp_data[CKB_SECP256K1_DATA_SIZE];
+  ret = ckb_secp256k1_custom_verify_only_initialize(&context, secp_data);
+  if (ret != 0) {
+    return ret;
+  }
+
+  size_t signature_offset = multisig_script_len;
+  for (size_t i = 0; i < threshold; i++) {
+    /* Load signature */
+    secp256k1_ecdsa_recoverable_signature signature;
+    signature_offset += i * SIGNATURE_SIZE;
+    if (secp256k1_ecdsa_recoverable_signature_parse_compact(
+            &context, &signature, &witness[signature_offset], witness[signature_offset + RECID_INDEX]) == 0) {
+      return ERROR_SECP_PARSE_SIGNATURE;
+    }
+
+    blake2b_state blake2b_ctx;
+    blake2b_init(&blake2b_ctx, BLAKE2B_BLOCK_SIZE);
+    blake2b_update(&blake2b_ctx, tx_hash, BLAKE2B_BLOCK_SIZE);
+    /* Load extrat witness of first input and all witnesses of other inputs */
+    if (extra_witness_len > 0) {
+      blake2b_update(&blake2b_ctx, &witness[required_witness_len], extra_witness_len);
+    }
+    size_t n = 1;
+    while (1) {
+      len = WITNESS_SIZE;
+      ret = ckb_load_witness(temp, &len, 0, n, CKB_SOURCE_GROUP_INPUT);
+      if (ret == CKB_INDEX_OUT_OF_BOUND) {
+        break;
+      }
+      if (ret != CKB_SUCCESS) {
+        return ERROR_SYSCALL;
+      }
+      if (len > WITNESS_SIZE) {
+        return ERROR_WITNESS_TOO_LONG;
+      }
+      blake2b_update(&blake2b_ctx, temp, len);
+      n += 1;
+    }
+    blake2b_final(&blake2b_ctx, temp, BLAKE2B_BLOCK_SIZE);
```

Comment:
The message to sign is identical for all signatures. You only need to compute it once.